app/lib/frontend/backend.dart

```diff
@@ -297,6 +296,38 @@ class Backend {
       await analyzerClient.triggerAnalysis(package, latestVersion, <String>{});
     });
   }
+
+  /// Whether the current authenticated user (or [userId] if provided) is a
+  /// package admin (through direct uploaders list or publisher admin).
+  ///
+  /// Returns false if the user is not an admin.
+  Future<bool> isPackageAdmin(models.Package p, {String userId}) async {
+    userId ??= authenticatedUser?.userId;
+    if (userId == null) {
+      return false;
+    }
+    if (p.publisherId == null) {
+      return p.containsUploader(userId);
+    } else {
+      final memberKey = db.emptyKey
+          .append(Publisher, id: p.publisherId)
+          .append(PublisherMember, id: userId);
+      final list = await db.lookup<PublisherMember>([memberKey]);
+      final member = list.single;
+      return member?.role == PublisherMemberRole.admin;
+    }
+  }
+
+  /// Whether the current authenticated user (or [userId] if provided) is a
+  /// package admin (through direct uploaders list or publisher admin).
+  ///
+  /// Throws exception if the user is not an admin.
+  Future checkPackageAdmin(models.Package package, {String userId}) async {
+    // Fail if calling user doesn't have permission to admin package.
```

Comment:
Also check first that the user is authenticated, if not throw the `AuthenticationMissingException`. Since `isPackageAdmin` won't check if the user is authenticated.

Reply:
it does check it:
- `authenticatedUser?.userId` is null if the user is not authenticated
- on the next line it returns if value is null

Reply:
Yes, but it doesn't throw `AuthenticationMissingException`

Reply:
My point was, we should throw an `AuthenticationMissingException` if th euser is not authenticated.

Throwing `AuthorizationException` when the user is not authenticated is always incorrect. How can we know if the user is authorized, if the user is not authenticated.